### PR TITLE
Global functions

### PR DESCRIPTION
**Pull request by CedNaru**
So far I implemented all the maths functions. Some are just wrappers for kotlin.math, some have to be reimplemented.
Some functions won't be implemented because they rely heavily on the dynamic nature of GdScript and it would be a pain to replicate that in Kotlin.
For now, I am waiting for the gen class in order to implement the random and resource loading functions. 

There is also a bunch of print functions inside Gdscript, I don't know if it's worth to implement them.

[Reference](https://docs.godotengine.org/en/3.2/classes/class_@gdscript.html)

### CODE REVIEW

File: godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            Variant.Type.DICTIONARY -> len(what.asDictionary())
+            Variant.Type.ARRAY -> len(what.asVariantArray())
+            Variant.Type.POOL_BYTE_ARRAY -> len(what.asPoolByteArray())
+            Variant.Type.POOL_INT_ARRAY -> len(what.asPoolIntArray())
+            Variant.Type.POOL_REAL_ARRAY -> len(what.asPoolRealArray())
+            Variant.Type.POOL_STRING_ARRAY -> len(what.asPoolStringArray())
+            Variant.Type.POOL_VECTOR2_ARRAY -> len(what.asPoolVector2Array())
+            Variant.Type.POOL_VECTOR3_ARRAY -> len(what.asPoolVector3Array())
+            Variant.Type.POOL_COLOR_ARRAY -> len(what.asPoolColorArray())
+            else -> throw IllegalArgumentException("Variant can not provide a length.")
```
Comment (chippmann) [inline godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt]: Shouldn't we in this case return -1?

Comment (CedNaru) [reply godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt]: The Gdscript doc says that it throws a fatal error when the variant type is not right. Just trying to be consistent with the original API.

Comment (chippmann) [reply godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt]: Ah ok. Then it's fine for me.

File: godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import godot.Object
+import godot.RandomNumberGenerator
+import godot.global.gdPrint
+import godot.global.gdResource
+import kotlinx.cinterop.invoke
+import kotlinx.cinterop.memScoped
+import kotlin.test.assertTrue
+
+
+object gd : gdMath, gdCore, gdRandom, gdPrint, gdResource {
```
Comment (chippmann) [inline godot-kotlin/godot-library/src/nativePublic/kotlin/godot/global/gd.kt]: Why is `gd` lowercase?

**Review by chippmann — APPROVED** [review]
LGTM. Thanks very much for the work! Especially the kDocs!
One question i have but nothing that prevents this from merging

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Quat.kt [hunk 3]
```diff
@@ -173,9 +173,9 @@ class Quat(var x: RealT, var y: RealT, var z: RealT, var w: RealT) : CoreType {
      */
     fun isEqualApprox(other: Quat): Boolean {
         return isEqualApprox(other.x, x)
-            && isEqualApprox(other.y, y)
-            && isEqualApprox(other.z, z)
-            && isEqualApprox(other.w, w)
+        isEqualApprox(other.y, y)
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Quat.kt]: I think it is an attention error. It is unreachable.

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Rect2.kt [hunk 4]
```diff
@@ -4,7 +4,7 @@ package godot.core
 
 import godot.gdnative.godot_rect2
 import godot.gdnative.godot_rect2_layout
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Rect2.kt:7]: Do we need the star import ? Or do we only need CoreType ?

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Transform.kt [hunk 5]
```diff
@@ -4,7 +4,7 @@ package godot.core
 
 import godot.gdnative.godot_transform
 import godot.gdnative.godot_transform_layout
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Transform.kt:7]: Same

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Transform2D.kt [hunk 6]
```diff
@@ -4,14 +4,15 @@ package godot.core
 
 import godot.gdnative.godot_transform2d
 import godot.gdnative.godot_transform2d_layout
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Transform2D.kt:7]: same

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Variant.kt [hunk 7]
```diff
@@ -5,13 +5,12 @@ import godot.Object
 import godot.gdnative.godot_array
 import godot.gdnative.godot_variant
 import godot.gdnative.godot_variant_type
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Variant.kt:7]: Same.

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/array/VariantArray.kt [hunk 8]
```diff
@@ -2,7 +2,7 @@ package godot.core
 
 import godot.Object
 import godot.gdnative.godot_array
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/array/VariantArray.kt:5]: Same.

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/array/core/CoreArray.kt [hunk 9]
```diff
@@ -2,7 +2,7 @@ package godot.core
 
 import godot.Object
 import godot.gdnative.godot_array
-import godot.internal.type.CoreType
+import godot.internal.type.*
```
Comment (piiertho) [inline godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/array/core/CoreArray.kt:5]: Same.

**Review by piiertho — CHANGES_REQUESTED** [review]
Few changes and comments, nothing to say on everything else :)
Thanks !

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Rect2.kt (hunk 4 above)
Comment (CedNaru) [reply godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Rect2.kt:7]: I changed this package. Now it contains the NaturalT/RealT alias and the isEqualApprox() function. Not putting the star and just optimizing the imports, add a ton of individual imports and  modify stuff like 
`godot.internal.type.isEqualApprox()`

File: godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Quat.kt (hunk 3 above)
Comment (CedNaru) [reply godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/Quat.kt]: Ha yeah. Like I say below. The optimized imported messed up that part of the code, I probably made a mistake when rewriting it.

**Review by piiertho — APPROVED** [review]
Thanks !

**Pull request merged**